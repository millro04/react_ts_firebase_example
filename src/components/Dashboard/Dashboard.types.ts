import { IBookData } from '../types';

export interface IDashboardState {
    bookData: IBookData[]
}

export interface IGraphData {
    data: number[],
    labels: string[],
    title: string
}

export interface IDashboardViewProps {
    ListGraphData: IGraphData[]
}
