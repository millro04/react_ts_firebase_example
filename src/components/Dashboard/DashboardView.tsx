import React from 'react';
import { Bar } from 'react-chartjs-2';
import { Grid, Row, Col } from 'react-flexbox-grid';
import './Dashboard.css';
import { IDashboardViewProps } from './Dashboard.types';

const chartOptions = {
  scales: {
    yAxes: [{
      scaleLabel: {
        display: true,
        labelString: 'Minutes'
      },
      ticks: {
        beginAtZero: true
    }
    }],
    xAxes: [{
      scaleLabel: {
        display: true,
        labelString: 'Date'
      }
    }]
  },
  maintainAspectRatio: false
}

export default class DashboardView extends React.Component<IDashboardViewProps, any> {

  renderGraph(labels: string[], data: number[], title: string) {
    const graphData: any = {
        labels: labels,
        datasets: [
          {
            label: title,
            backgroundColor: 'rgba(75,192,192,0.4)',
            borderColor: 'rgba(75,192,192,1)',
            borderWidth: 1,
            hoverBackgroundColor: 'rgba(75,192,192,0.6)',
            hoverBorderColor: 'rgba(75,192,192,1)',
            data: data
          }
        ]
      };
      return (
        <Col className='graph' xs={6} sm={6} md={6} lg={6}>
            <Bar
                data={graphData}
                height={300}
                options={chartOptions} />
        </Col>
      );
  }

  render() {
    return (
      <div style={{ paddingTop: '15px' }}>
          <Grid fluid>
              <Row>
                {this.props.ListGraphData.map(graphData => 
                    (this.renderGraph(graphData.labels, graphData.data, graphData.title))
                )}
              </Row>
          </Grid>
        
      </div>
    );
  }
}