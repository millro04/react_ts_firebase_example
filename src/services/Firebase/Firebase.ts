import * as firebase from 'firebase';

// TODO: Would not hard-code these for production.
const config = {
    apiKey: "AIzaSyAJyGhmH2K9QNvYwFu16JVwse8JzN5A0ps",
    authDomain: "readingtracker-ed2b9.firebaseapp.com",
    databaseURL: "https://readingtracker-ed2b9.firebaseio.com",
    projectId: "readingtracker-ed2b9",
    storageBucket: "readingtracker-ed2b9.appspot.com",
    messagingSenderId: "5558996582",
    appId: "1:5558996582:web:6dcc494fca9c321f6be7a7",
    measurementId: "G-4WCDW2XQYE"
};
firebase.initializeApp(config);

firebase.firestore();

export default firebase;