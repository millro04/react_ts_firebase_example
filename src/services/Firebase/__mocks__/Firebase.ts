// Sample Firebase mock used for testing
// See: https://stackoverflow.com/questions/52043886/how-do-you-mock-firebase-firestore-methods-using-jest

const docData = { data: "MOCK_DATA" };
const docResult = {
  // simulate firestore get doc.data() function
  data: () => docData
};
const get = jest.fn(() => Promise.resolve(docResult));
const firebaseSet = jest.fn(() => Promise.resolve());
const firebaseAdd = jest.fn(() => Promise.resolve());
const firebaseDelete = jest.fn(() => Promise.resolve());
const onSnapshot = jest.fn();

const doc = jest.fn(() => {
  return {
    set: firebaseSet,
    get,
    delete: firebaseDelete
  };
});

const collection = () => {
  return {
    onSnapshot,
    add: firebaseAdd,
    doc
  }
};

const firestore = () => {
  return { collection };
};

export { firestore, firebaseAdd, firebaseDelete, firebaseSet, onSnapshot };