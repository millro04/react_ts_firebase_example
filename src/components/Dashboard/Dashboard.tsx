import React from 'react';
import DashboardView from './DashboardView';
import firebase from '../../services/Firebase/Firebase';

interface IBookData {
    date: string,
    book_name: string,
    min_read: number, 
}

interface IDashboardState {
    bookData: IBookData[]
}

export default class Dashboard extends React.Component<any, IDashboardState> {

  firebaseRef: any;
  constructor(props: React.ReactPropTypes) {
      super(props);
      this.firebaseRef = firebase.firestore().collection('reading_logs');
      this.state = {
          bookData: []
      }
  }

  componentDidMount() {
    this.firebaseRef.onSnapshot(this.onReadingCollectionUpdate);
  }

  onReadingCollectionUpdate = (querySnapshot: any) => {
    const bookData: any = [];
    querySnapshot.forEach((doc: any) => {
      const { book_name, date_read, minutes_read } = doc.data();
      bookData.push(
        {
          date: date_read,
          book_name: book_name,
          min_read: minutes_read,
          id: doc.id,
        });
        this.setState({
          bookData
       });
    });
  }

  formatBookDataForGraphs() {
    let resultsObj: any = {};
    let resultsArray: any = [];
    this.state.bookData.forEach(
        elem => {
            if (resultsObj.hasOwnProperty(elem.book_name)) {
                resultsObj[elem.book_name].labels.push(elem.date);
                resultsObj[elem.book_name].data.push(elem.min_read);
            } else {
                resultsObj[elem.book_name] = {labels: [elem.date], data: [elem.min_read], title: `Minutes Read Per Day - ${elem.book_name}`}
            }
        });
    Object.keys(resultsObj).forEach(function(key) {
        resultsArray.push(
            {
                title: resultsObj[key].title,
                data: resultsObj[key].data,
                labels: resultsObj[key].labels,

            }
        )
    });
    return resultsArray;
  }

  render() {
    return (
      <DashboardView
        ListGraphData={this.formatBookDataForGraphs()}
        />
    );
  }
}