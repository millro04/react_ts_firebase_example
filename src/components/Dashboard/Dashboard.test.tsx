import React from 'react';
import Dashboard from './Dashboard';
import DashboardView from './DashboardView';
import { shallow, ShallowWrapper, configure, ReactWrapper, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({adapter: new Adapter()});
import firebase from '../../services/Firebase/Firebase';
import { firestore } from '../../services/Firebase/__mocks__/Firebase';
firebase.firestore = firestore;

test('Dashboard component renders', () => {
    const component: ShallowWrapper = shallow(<Dashboard />);
    expect(component).toMatchSnapshot();
});

test('DashboardView component renders', () => {
    const component: ShallowWrapper = shallow(<DashboardView ListGraphData={[]} />);
    expect(component).toMatchSnapshot();
});

