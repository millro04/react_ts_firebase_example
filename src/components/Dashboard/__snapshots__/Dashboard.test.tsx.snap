// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard component renders 1`] = `ShallowWrapper {}`;

exports[`DashboardView component renders 1`] = `ShallowWrapper {}`;
